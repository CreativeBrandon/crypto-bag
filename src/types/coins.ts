export interface CoinsEntities {
    loaded: boolean
    isLoading: boolean
    ids: string[]
    byIds: ByIds
}

export interface CoinsResponse {
    data: Coin[]
}

export interface Coin {
    id: string
    name: string
    symbol: string
    rank: number
    price_usd: number
    price_btc: number
    '24h_volume_usd': number
    market_cap_usd: number
    available_supply: number
    total_supply: number
    max_supply: number
    percent_change_1h: number
    percent_change_24h: number
    percent_change_7d: number
    last_updated: number
}

export interface ByIds {
    [index: string]: Coin
}
