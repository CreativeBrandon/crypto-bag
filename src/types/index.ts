export * from './bags'
export * from './coins'
export * from './store.d'
