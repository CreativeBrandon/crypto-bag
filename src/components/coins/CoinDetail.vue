<template>
    <div class="coin-detail">
        <img class="icon" :src="iconPath" :alt="coin.name" /> 
        <h3 class="name">{{coin.rank}} {{coin.name}}</h3>
        <div><span>Price:</span> {{coin.price_usd | currency}} USD</div>
        <div><span>Sat:</span> {{coin.price_btc}} /BTC</div>
        <div><span>Market Cap:</span> {{coin.market_cap_usd | currency}}</div>
    </div>
</template>

<style scoped >
.coin-detail {
  margin-bottom: 20px;
}
.icon {
  height: 32px;
  max-width: 100px;
}
span {
  font-weight: 600;
}
</style>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { Coin } from '@/types/coins'

@Component
export default class CoinDetail extends Vue {
  @Prop() coin?: Coin
  iconPath: string = ''

  created() {
    const icon = this.coin
      ? this.coin.symbol.toLowerCase().replace(/[0-9]/g, 'X')
      : null
    if (icon) this.setIconPath(icon)
  }

  setIconPath(icon: string) {
    const iconPath = require(`@creativebrandon/cryptocurrency-icons/svg/color/${icon}.svg`)
    if (iconPath) return (this.iconPath = iconPath)
  }
}
</script>
