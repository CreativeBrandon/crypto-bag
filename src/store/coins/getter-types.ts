import { CoinState as State } from '@/types'

export const GET_COINS = '[COINS] GET COINS'
export type GET_COINS = typeof GET_COINS

export const IS_LOADING = '[COINS] IS LOADING?'
export type IS_LOADING = typeof IS_LOADING
