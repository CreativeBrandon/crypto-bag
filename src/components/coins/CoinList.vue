<template>
    <section class="coin-list">
        <ul>
            <coinDetail 
                v-for="coin in coins"
                :coin="coin"
                :key="coin.id" />
        </ul>
    </section>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { Coin } from '@/types/coins'

const coinDetail = () => import('./CoinDetail.vue')

@Component({
    components: { coinDetail }
})
export default class CoinList extends Vue {
    @Prop() coins?: Coin[]
}
</script>

<style scoped>
.coin-list {
    margin: 20px 0;
}
</style>
