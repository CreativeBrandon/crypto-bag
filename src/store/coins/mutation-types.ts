import { State } from './store'
import * as utils from '@/utils'
import { Coin } from '@/types/coins'

export const FETCH_COINS_SUCCESS = '[COINS] FETCH COINS SUCCESS'
export type FETCH_COINS_SUCCESS = typeof FETCH_COINS_SUCCESS

export const FETCH_COINS_ERROR = '[COINS] FETCH COINS ERROR'
export type FETCH_COINS_ERROR = typeof FETCH_COINS_ERROR

export const LOADING = '[COINS] LOADING'
export type LOADING = typeof LOADING

export const fetchCoinsSuccess = (state: State, payload: Coin[]) => {
    state.entities = {
        ...state.entities,
        isLoading: false,
        loaded: true,
        ids: [
            ...state.entities.ids,
            ...payload.map((coin: Coin) => coin.id)
        ],
        byIds: {
            ...state.entities.byIds,
            ...utils.objFromArray(payload)
        }
    }
}

export const fetchCoinsError = (state: State, payload: any) => {
    state.entities.loaded = false
    state.errors = payload
}
