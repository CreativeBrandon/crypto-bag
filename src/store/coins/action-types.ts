import { ActionContext } from 'vuex'
import { State } from './store'
import * as mutation from './mutation-types'
import { coinsService } from '@/services/coins.service'

// ACTIONS
export const FETCH_COINS = '[COINS] FETCH_COINS'
export type FETCH_COINS = typeof FETCH_COINS

export const FETCH_COINS_SUCCESS = '[COINS] FETCH_COINS_SUCCESS'
export type FETCH_COINS_SUCCESS = typeof FETCH_COINS_SUCCESS

export const FETCH_COINS_BY_CATEGORY = '[COINS] FETCH_COINS_BY_CATEGORY'
export type FETCH_COINS_BY_CATEGORY = typeof FETCH_COINS_BY_CATEGORY

export const fetchCoins = (context: ActionContext<State, any>) => {
    if (!context.state.entities.loaded) {
        context.commit(mutation.LOADING)
        coinsService
            .fetchCoins(25)
            .then(res => context.commit(mutation.FETCH_COINS_SUCCESS, res.data))
            .catch(err => context.commit(mutation.FETCH_COINS_ERROR, err))
    }
}
