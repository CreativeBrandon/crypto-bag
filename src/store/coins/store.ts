import Vue from 'vue'
import Vuex, { ActionTree, GetterTree, MutationTree } from 'vuex'
import { CoinSearch, CoinState as State } from '@/types'
import * as action from './action-types'
import * as getter from './getter-types'
import * as mutation from './mutation-types'

export const mutations: MutationTree<State> = {
    [mutation.FETCH_COINS_SUCCESS]: mutation.fetchCoinsSuccess,
    [mutation.FETCH_COINS_ERROR]: mutation.fetchCoinsError,
    [mutation.LOADING]: (state: State) => state.entities.isLoading = true,
    [mutation.SEARCH_RESULTS]: (state: State, payload: CoinSearch) => state.search = payload
}

// any = RootState
export const getters: GetterTree<State, any> = {
    [getter.GET_COINS]: state => state.entities.ids.map(id => state.entities.byIds[id]),
    [getter.IS_LOADING]: state => state.entities.isLoading
}

export const actions: ActionTree<State, any> = {
    [action.FETCH_COINS]: action.fetchCoins,
    [action.SEARCH_COINS]: action.searchCoins
}

export const state = (): State => {
    return {
      errors: {},
      search: {
          query: '',
          results: []
      },
      entities: {
        loaded: false,
        isLoading: false,
        ids: [],
        byIds: {}
      }
    }
}

export const store = {
    namespace: true,
    state: state(),
    mutations,
    actions,
    getters
}

export default store
