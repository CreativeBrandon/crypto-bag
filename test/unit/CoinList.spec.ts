import Vue from 'vue'
import { shallow } from '@vue/test-utils'
import { Coin, coinMock } from '@/types/coins'
import coinList from '@/components/coins/coinList.vue'

const factory = (props = {}) => {
    return shallow(coinList, {
        propsData: { ...props }
    })
}

describe('coin list Component', () => {
    const coin = coinMock() as Coin
    const wrapper = factory({ coins: [coin] })

    test('hasclass coin-detail', () => {
        expect(wrapper.find('.coin-list').exists()).toBe(true)
    })
})
