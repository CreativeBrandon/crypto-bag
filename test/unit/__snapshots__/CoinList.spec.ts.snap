// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coin list Component renders a list of coin detail components 1`] = `[Function]`;
